src/biom_inmem.hpp

```diff
@@ -0,0 +1,169 @@
+/*
+ * BSD 3-Clause License
+ *
+ * Copyright (c) 2016-2023, UniFrac development team.
+ * All rights reserved.
+ *
+ * See LICENSE file for more details
+ */
+
+
+#ifndef _UNIFRAC_BIOM_INMEM_H
+#define _UNIFRAC_BIOM_INMEM_H
+
+#include <vector>
+#include <unordered_map>
+
+#include "biom_interface.hpp"
+
+namespace su {
+    class sparse_data {
+        public:
+            /* default constructor */
+            sparse_data(bool _clean_on_destruction);
+
+            /* constructor from compress sparse data
+             *
+             * @param n_obs number of observations
+             * @param index vector of index positions
+             * @param indptr vector of indptr positions
+             * @param data vector of observation counts
+             */
+            sparse_data(const uint32_t n_obs,
+                        const uint32_t n_samples,
+                        uint32_t* index,
+                        uint32_t* indptr,
+                        double* data);
+
+            /* default destructor */
+            virtual ~sparse_data();
+
+            /* modified copy constructor */
+            sparse_data(const sparse_data& other, bool _clean_on_destruction);
+
+            /* remove ownership of this obs_idx buffer */
+            uint32_t *steal_indeces(uint32_t obs_idx) {uint32_t *out=obs_indices_resident[obs_idx]; obs_indices_resident[obs_idx]=NULL; return out;}
```

Comment:
Minor but I think this should be `steal_indices`

Reply:
Good point, renamed.